write-fonts/src/tables/cmap.rs

```diff
@@ -237,6 +236,187 @@ impl Cmap {
+                }
+            // and the funny case:
+            // if we were not previously contiguous but are now:
+            // - if i == 0, then this is the first item in a new segment;
+            //   set is_contiguous and continue
+            // - if i > 0, we need to back up one
+            } else if !self.currently_contiguous {
+                if i == 0 {
+                    self.currently_contiguous = true;
+                } else {
+                    return Some(self.make_segment(i - 1));
+                }
+            }
+            prev_cp = *cp;
+            prev_gid = *gid;
+        }
+
+        // if we're done looping then create the last segment:
+        let last_idx = self.mappings.len() - 1;
+        Some(self.make_segment(last_idx - self.seg_start))
+    }
+}
+
+/// Computes an efficient set of segments
+fn compute_format_4_segments(mappings: &[(char, GlyphId)]) -> Vec<Format4Segment> {
+    assert!(!mappings.is_empty());
+    let mut iter = Format4Ranges::new(mappings).peekable();
+    let Some(first) = iter.next() else {
+        return Default::default();
+    };
+
+    let mut result = vec![first];
+
+    // now we want to collect the segments, combining smaller segments where
+    // that leads to a size savings.
+    //
+    // This differs from the python, which starts from larger segments and then
+    // subdivides them, but the overall idea is the same.
+    // (https://github.com/fonttools/fonttools/blob/f1d3e116d54f/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L783)
+    while let Some(segment) = iter.next() {
+        let prev = result.last_mut().unwrap();
+        let is_contiguous_with_prev =
+            mappings[prev.end_ix].0 as u32 + 1 == mappings[segment.start_ix].0 as u32;
+        let is_contiguous_with_next = iter
+            .peek()
+            .map(|next| mappings[segment.end_ix].0 as u32 + 1 == mappings[next.start_ix].0 as u32)
+            .unwrap_or(false);
+        // first: if segment is not contiguous with either prev or next, it can't be
+        // combined, so just push and continue
+
+        if !(is_contiguous_with_prev | is_contiguous_with_next) {
+            result.push(segment);
+            continue;
+        }
+
+        // next: if chars are contiguous and neither has a delta, we always combine
+        // this will mostly happen if we've combined the previous, previously
+        // delta-having segment
+        if is_contiguous_with_prev && prev.id_delta.is_none() && segment.id_delta.is_none() {
+            *prev = prev.combine(segment);
+            continue;
+        }
+        // next: if contiguous, combine only if it saves bytes
+        if is_contiguous_with_prev {
+            let combined = prev.combine(segment);
+            if combined.cost() < prev.cost() + segment.cost() {
+                *prev = combined;
+                continue;
+            }
+        }
+        result.push(segment);
+    }
+    result
+}
```

Comment:
I'm slightly surprised the loop doesn't mirror the comment more directly, e.g. see if we can combine, do so if it saves bytes? - naive pseudocode:

```rust
while let Some(next) = iter.next() {
  // we initialized result with the first result from iter
  if !result.last_mut().unwrap().try_combine(next) {
    result.push(next);
  }
}
```

Checking next and prev and checking id delta instead of letting the cost check handle it both throw me.

Reply:
https://github.com/googlefonts/fontations/pull/1230 tries ^ and seems to pass the tests.

Reply:
yea I think this is a nice idea, will play around a bit.